File: src/lib/crawl/storage.ts
// ============================================================
// 存储系统 — 章节TXT文件 / 封面webp转换 / 目录管理
// 章节可写库(SQLite)或生成txt文件存指定文件夹
// 封面统一下载转为 webp 格式存储
// ============================================================
import { promises as fs } from 'fs'
import path from 'path'
import sharp from 'sharp'

export const DATA_ROOT = path.join(process.cwd(), 'data')
export const NOVELS_DIR = path.join(DATA_ROOT, 'novels')
export const COVERS_DIR = path.join(DATA_ROOT, 'covers')
export const DOWNLOADS_DIR = path.join(DATA_ROOT, 'downloads')

export async function ensureDirs() {
  await Promise.all([
    fs.mkdir(NOVELS_DIR, { recursive: true }),
    fs.mkdir(COVERS_DIR, { recursive: true }),
    fs.mkdir(DOWNLOADS_DIR, { recursive: true }),
  ])
}

/** 章节txt存储: data/novels/{bookId}/{idx pad5}_{slug}.txt */
export async function saveChapterTxt(
  bookId: string,
  idx: number,
  title: string,
  content: string
): Promise<string> {
  await ensureDirs()
  // R26-1A P1 修复: bookId 路径穿越防御。bookId 通常为 Prisma cuid() (字母数字,
  // 25 字符固定形态), 但防御性 Coding: 剥除所有路径分隔符与父目录指针字符后
  // 才拼路径, 防 caller 侧误传 / DB 迁移 / 手工调用传入非法 ID 造成越出 NOVELS_DIR。
  // 同口径于 slug 处的剥除集, 但额外剥 ' . \0 与路径分隔符(Caddy 不会转发这些字符,
  // 但防御在源头)
  const safeBookId = String(bookId || '').replace(/[\\/\x00\s.]+/g, '_').replace(/^_+|_+$/g, '') || 'unknown_book'
  const dir = path.join(NOVELS_DIR, safeBookId)
  await fs.mkdir(dir, { recursive: true })
  // 清洗控制字符(\x00-\x1f, 原 \s 不覆盖) + Windows 保留字符; 截断防超长文件名
  // (Bug 21 修复: 按码点截断: 直接 slice(0,80) 可能把 emoji 等 astral 字符的代理对拦腰
  // 斩断, 落盘出半字符乱码文件名; Array.from 按码点迭代后再 slice 才安全)
  const slug = Array.from(title.replace(/[\x00-\x1f\\/:*?"<>|\s]+/g, '_')).slice(0, 80).join('')
  const slugSafe = Array.from(slug).slice(0, 40).join('') || 'chapter'
  const fileName = `${String(idx).padStart(5, '0')}_${slugSafe}.txt`
  const filePath = path.join(dir, fileName)
  // R3-27: 标题强制单行 —— 源站标题偶含 \r\n(列表项跨行/HTML br 转文本时残留), 写入
  // "${title}\n\n${content}\n" 后会被分割成多行, 读取侧 readChapterTxt.split('\n').slice(1)
  // 会把标题尾行误当正文首段。落盘前剥成单行(替换 \r\n 为单空格)
  const safeTitle = title.replace(/[\r\n]+/g, ' ')
  // R26-1A P1 修复: 原子写入。原实现 fs.writeFile(filePath, ...) 非原子: 进程崩溃
  // / 磁盘满 / 写入中被信号中断 → 产出半成品文件被后续读取者读到损坏内容(公共
  // /api/public/read 章节接口直接读 filePath 返正文)。改为先写 .tmp 临时文件再 rename。
  // fs.rename 在 POSIX 同文件系统上是原子的(原子 inode 替换), 中间任何时点读者
  // 看到的都是【旧完整文件】或【新完整文件】之一, 永远看不到半成品。跨设备 rename
  // 会 fallback 到 copy+unlink(非原子), 但本场景 filePath 与 .tmp 同目录同设备, 安全。
  // 临时文件名加 PID + 随机段防并发同章节写入互踩(同一本书同章节并发采集场景)
  const tmpPath = `${filePath}.${process.pid}.${Math.floor(Math.random() * 1e9)}.tmp`
  try {
    await fs.writeFile(tmpPath, `${safeTitle}\n\n${content}\n`, 'utf-8')
    await fs.rename(tmpPath, filePath)
  } catch (e) {
    // 临时文件清理: rename 成功后 tmpPath 不存在, unlink 静默失败; rename 失败时
    // tmpPath 可能存在半成品(磁盘满/权限), 主动清避免泄漏(.tmp 文件不会被公共
    // API 读取, 但占用磁盘空间)
    try { await fs.unlink(tmpPath) } catch { /* tmp 不存在或已被 rename, 忽略 */ }
    throw e
  }
  return path.relative(DATA_ROOT, filePath) // 相对 data/ 的路径
}

export async function readChapterTxt(relPath: string): Promise<string | null> {
  try {
    const full = path.resolve(DATA_ROOT, relPath)
    // 修复: startsWith(DATA_ROOT) 存在同级目录前缀绕过(data vs data-covers),
    // 必须以 path.sep 结尾的前缀匹配才算落在数据目录内
    if (full !== DATA_ROOT && !full.startsWith(DATA_ROOT + path.sep)) return null
    return await fs.readFile(full, 'utf-8')
  } catch {
    return null
  }
}

export async function deleteBookTxt(bookId: string) {
  try {
    await fs.rm(path.join(NOVELS_DIR, bookId), { recursive: true, force: true })
  } catch { /* ignore */ }
}

/** 封面下载 → webp (sharp), 返回相对路径 covers/{name}.webp
 *  转换失败时降级: 先用宽容模式重试, 再失败则回存原始图片字节(公开封面接口按 .webp
 *  文件名提供服务, 浏览器 <img> 解码时按魔数嗅探实际格式, 不影响展示) */
export async function saveCoverWebp(
  buf: Buffer,
  name: string
): Promise<string | null> {
  try {
    await ensureDirs()
    if (!buf || buf.length === 0 || buf.length > 20 * 1024 * 1024) return null // 空文件/超大文件保护
    const safeName = name.replace(/[^\w-]/g, '') || `cover_${Date.now()}_${Math.floor(Math.random() * 9999)}`
    const fileName = `${safeName}.webp`
    const filePath = path.join(COVERS_DIR, fileName)
    try {
      await sharp(buf)
        .resize(400, 533, { fit: 'inside', withoutEnlargement: true })
        .webp({ quality: 82 })
        .toFile(filePath)
    } catch {
      // 降级1: 宽容模式(容忍截断/轻微损坏的图)
      try {
        await sharp(buf, { failOn: 'none' })
          .resize(400, 533, { fit: 'inside', withoutEnlargement: true })
          .webp({ quality: 82 })
          .toFile(filePath)
      } catch {
        // 降级2: 回存原始字节(保持 .webp 文件名以兼容封面接口的文件名白名单)
        await fs.writeFile(filePath, buf)
      }
    }
    return `covers/${fileName}`
  } catch (e) {
    console.warn('[storage] cover webp convert failed', e)
    return null
  }
}

export async function readCover(fileName: string): Promise<Buffer | null> {
  try {
    const safe = path.basename(fileName)
    const full = path.join(COVERS_DIR, safe)
    if (!full.startsWith(COVERS_DIR)) return null
    return await fs.readFile(full)
  } catch {
    return null
  }
}

/** 下载成品文件名计算(openDownloadTxtWriter 共用口径):
 *  清洗控制字符 + 截断: 超长书名会导致 ENAMETOOLONG 直接抛错(按码点截断防代理对斩半) */
function downloadTxtTarget(name: string): { filePath: string; rel: string; fileName: string } {
  // Bug 21 修复: 原实现 .slice(0, 100) 按 UTF-16 code unit 截断, astral 字符(emoji/CJK 扩展)
  // 代理对被斩半 → 落盘出半字符乱码文件名。改为 Array.from 按码点迭代后再 slice。
  const base = Array.from(name.replace(/[\x00-\x1f\\/:*?"<>|\s]+/g, '_')).slice(0, 100).join('')
  const fileName = `${Array.from(base).slice(0, 80).join('')}.txt`
  const filePath = path.join(DOWNLOADS_DIR, fileName)
  return { filePath, rel: `downloads/${fileName}`, fileName }
}

/**
 * 下载成品流式写入器(gg-a): 逐段 append 写盘, 全书不再内存拼接后单次落盘 ——
 * 万章书的数百 MB 级 parts 数组峰值(原实现)对大书是生成即 OOM 的形态。
 * 用法: write(header/章节段…) → finish() 落定返回 {rel,size}; 中途失败 abort()
 * 删除半成品(与原实现"失败即无文件"卫生语义一致)。
 * 字节等价性: UTF-8 按码点编码无跨段状态, 分段写入与整串写入逐字节一致
 * (verify-gg-a-txt-stream 以修前基准 diff 实证)
 */
export interface DownloadTxtWriter {
  rel: string
  write(chunk: string): Promise<void>
  finish(): Promise<{ rel: string; size: number }>
  abort(): Promise<void>
}

export async function openDownloadTxtWriter(name: string): Promise<DownloadTxtWriter> {
  await ensureDirs()
  const { filePath, rel } = downloadTxtTarget(name)
  const fh = await fs.open(filePath, 'w')
  return {
    rel,
    async write(chunk: string) {
      if (!chunk) return
      await fh.write(chunk, undefined, 'utf-8')
    },
    async finish() {
      await fh.close()
      const stat = await fs.stat(filePath)
      return { rel, size: stat.size }
    },
    async abort() {
      try { await fh.close() } catch { /* 已关闭忽略 */ }
      try { await fs.unlink(filePath) } catch { /* ignore */ }
    },
  }
}

File: go-backend/crawl/storage.go
// storage.go — 章节 TXT / 封面 webp / 路径穿越防御 + 原子写入 (R38-1C).
//
// 与 TS 端 src/lib/crawl/storage.ts 同口径核心功能:
//   - DATA_ROOT / NOVELS_DIR / COVERS_DIR / DOWNLOADS_DIR (基于 CWD)
//   - EnsureDirs (MkdirAll 三个目录)
//   - SaveChapterTxt (bookId 路径穿越防御 + 标题 slug 清洗 + 原子写入 .tmp+rename)
//   - ReadChapterTxt (sibling-prefix 绕过防御 + path.sep 结尾前缀匹配)
//   - DeleteBookTxt (bookId 路径穿越防御 + 同款清洗)
//   - SaveCoverWebp (封面字节存 .webp; Go 端无 sharp, 直接回存原始字节, 浏览器按魔数嗅探)
//   - ReadCover (sibling-prefix 绕过防御 + path.basename 剥目录组件)
//   - OpenDownloadTxtWriter (流式写入器, 万章书不 OOM)
package crawl

import (
        "context"
        "crypto/rand"
        "encoding/binary"
        "fmt"
        "io"
        "os"
        "path/filepath"
        "regexp"
        "strings"
        "sync"
        "time"
)

var (
        // 数据目录基于 CWD (与 main.go basePath 同口径)
        storageOnce sync.Once
        dataRoot    string
        novelsDir   string
        coversDir   string
        downloadsDir string
)

func initStoragePaths() {
        storageOnce.Do(func() {
                cwd, _ := os.Getwd()
                // 找项目根 (go-backend 的上级)
                root := cwd
                if _, err := os.Stat(filepath.Join(cwd, "go-backend")); err == nil {
                        root = cwd
                } else if strings.HasSuffix(cwd, "go-backend") {
                        root = filepath.Dir(cwd)
                }
                dataRoot = filepath.Join(root, "data")
                novelsDir = filepath.Join(dataRoot, "novels")
                coversDir = filepath.Join(dataRoot, "covers")
                downloadsDir = filepath.Join(dataRoot, "downloads")
        })
}

// DataRoot — 数据根目录绝对路径.
func DataRoot() string {
        initStoragePaths()
        return dataRoot
}

// NovelsDir — 章节TXT存储目录.
func NovelsDir() string {
        initStoragePaths()
        return novelsDir
}

// CoversDir — 封面存储目录.
func CoversDir() string {
        initStoragePaths()
        return coversDir
}

// DownloadsDir — 下载成品存储目录.
func DownloadsDir() string {
        initStoragePaths()
        return downloadsDir
}

// EnsureDirs — 确保数据目录存在.
func EnsureDirs() error {
        initStoragePaths()
        for _, d := range []string{novelsDir, coversDir, downloadsDir} {
                if err := os.MkdirAll(d, 0755); err != nil {
                        return err
                }
        }
        return nil
}

// ---------- 路径清洗工具 ----------

var (
        safeBookIdRe   = regexp.MustCompile(`[\\/\x00\s.]+`)
        trimUnderscore = regexp.MustCompile(`^_+|_+$`)
        chapterSlugRe  = regexp.MustCompile(`[\x00-\x1f\\/:*?"<>|\s]+`)
        coverNameRe    = regexp.MustCompile(`[^\w-]`)
)

// sanitizeBookId — bookId 路径穿越防御: 剥所有路径分隔符与父目录指针字符后 + 去首尾 _.
// 空串兜底 'unknown_book'.
func sanitizeBookId(bookID string) string {
        s := safeBookIdRe.ReplaceAllString(bookID, "_")
        s = trimUnderscore.ReplaceAllString(s, "")
        if s == "" {
                return "unknown_book"
        }
        return s
}

// sanitizeChapterSlug — 章节标题 slug 清洗: 控制字符 + Windows 保留字符 + 空白 → _.
// 按码点截断防代理对斩半 (Bug 21 修复).
func sanitizeChapterSlug(title string, maxRunes int) string {
        if maxRunes <= 0 {
                maxRunes = 80
        }
        cleaned := chapterSlugRe.ReplaceAllString(title, "_")
        runes := []rune(cleaned)
        if len(runes) > maxRunes {
                runes = runes[:maxRunes]
        }
        return string(runes)
}

// sanitizeCoverName — 封面文件名清洗: 仅保留字母数字和 -.
func sanitizeCoverName(name string) string {
        cleaned := coverNameRe.ReplaceAllString(name, "")
        return cleaned
}

// ---------- SaveChapterTxt ----------

// SaveChapterTxt — 章节txt存储: data/novels/{bookId}/{idx pad5}_{slug}.txt
//  - bookId 路径穿越防御 (剥路径分隔符 + 父目录指针字符)
//  - 标题 slug 清洗 (控制字符 + Windows 保留字符 + 按码点截断)
//  - 标题强制单行 (剥 \r\n → 单空格, 防 readChapterTxt.split('\n').slice(1) 误把标题尾行当正文首段)
//  - 原子写入: 先写 .tmp + os.Rename (POSIX 同文件系统原子 inode 替换)
//  - 临时文件名加 PID + 随机段防并发同章节写入互踩
//  返回相对 data/ 的路径 (供 DB 存储 + 公共 read API 使用).
func SaveChapterTxt(bookID string, idx int, title, content string) (string, error) {
        if err := EnsureDirs(); err != nil {
                return "", err
        }
        safeBookID := sanitizeBookId(bookID)
        dir := filepath.Join(novelsDir, safeBookID)
        if err := os.MkdirAll(dir, 0755); err != nil {
                return "", err
        }
        slug := sanitizeChapterSlug(title, 80)
        if slug == "" {
                slug = "chapter"
        }
        slugSafe := sanitizeChapterSlug(slug, 40)
        if slugSafe == "" {
                slugSafe = "chapter"
        }
        fileName := fmt.Sprintf("%05d_%s.txt", idx, slugSafe)
        filePath := filepath.Join(dir, fileName)
        // 标题强制单行 (源站标题偶含 \r\n, 落盘前剥成单行)
        safeTitle := strings.ReplaceAll(title, "\r", " ")
        safeTitle = strings.ReplaceAll(safeTitle, "\n", " ")
        body := safeTitle + "\n\n" + content + "\n"
        // 临时文件名: PID + 随机段
        var randBuf [8]byte
        _, _ = rand.Read(randBuf[:])
        randNum := binary.LittleEndian.Uint64(randBuf[:])
        tmpPath := fmt.Sprintf("%s.%d.%d.tmp", filePath, os.Getpid(), randNum)
        if err := os.WriteFile(tmpPath, []byte(body), 0644); err != nil {
                _ = os.Remove(tmpPath)
                return "", err
        }
        if err := os.Rename(tmpPath, filePath); err != nil {
                _ = os.Remove(tmpPath)
                return "", err
        }
        // 相对 data/ 的路径
        rel, err := filepath.Rel(dataRoot, filePath)
        if err != nil {
                return "", err
        }
        // 统一为 unix 风格 (DB 存储口径)
        rel = filepath.ToSlash(rel)
        return rel, nil
}

// ---------- ReadChapterTxt ----------

// ReadChapterTxt — 读取章节TXT (相对 data/ 的路径).
//  - 路径穿越防御: 必须 === DATA_ROOT 或以 DATA_ROOT + sep 开头
//    (防 sibling-prefix 绕过: data vs data-covers)
//  - 不存在/越界 → 返回 ("", nil)
func ReadChapterTxt(relPath string) (string, error) {
        initStoragePaths()
        // filepath.Join 已 Clean (剥 ..), 但仍需 sibling-prefix 防御: 直接拼 + 前缀校验
        full := filepath.Join(dataRoot, filepath.Clean(filepath.ToSlash(relPath)))
        if full == dataRoot {
                return "", nil
        }
        if !strings.HasPrefix(full, dataRoot+string(filepath.Separator)) {
                return "", nil
        }
        data, err := os.ReadFile(full)
        if err != nil {
                if os.IsNotExist(err) {
                        return "", nil
                }
                return "", err
        }
        return string(data), nil
}

// ---------- DeleteBookTxt ----------

// DeleteBookTxt — 删除整本书的 TXT 目录 (bookId 路径穿越防御).
//  - 与 SaveChapterTxt 同款清洗后再拼路径, 防 caller 误传 '../../etc' 等恶意 ID.
func DeleteBookTxt(bookID string) error {
        safeBookID := sanitizeBookId(bookID)
        dir := filepath.Join(novelsDir, safeBookID)
        return os.RemoveAll(dir)
}

// ---------- SaveCoverWebp ----------

// SaveCoverWebp — 封面字节存 .webp. Go 端无 sharp 库, 直接回存原始字节.
//  - 公开封面接口按 .webp 文件名提供服务, 浏览器 <img> 解码时按魔数嗅探实际格式,
//    不影响展示 (与 TS 端 saveCoverWebp 降级2 回存原始字节同口径)
//  - 空文件/超大文件保护 (>20MB 拒绝)
//  - 文件名: 仅保留 [\w-] 字符, 空串兜底 cover_{ts}_{rand}
//  返回相对 data/ 的路径 (covers/{name}.webp)
func SaveCoverWebp(buf []byte, name string) (string, error) {
        if len(buf) == 0 || len(buf) > 20*1024*1024 {
                return "", nil
        }
        if err := EnsureDirs(); err != nil {
                return "", err
        }
        safeName := sanitizeCoverName(name)
        if safeName == "" {
                var randBuf [4]byte
                _, _ = rand.Read(randBuf[:])
                randNum := binary.LittleEndian.Uint32(randBuf[:])
                safeName = fmt.Sprintf("cover_%d_%d", time.Now().UnixMilli(), randNum)
        }
        fileName := safeName + ".webp"
        filePath := filepath.Join(coversDir, fileName)
        if err := os.WriteFile(filePath, buf, 0644); err != nil {
                return "", err
        }
        return "covers/" + fileName, nil
}

// ---------- ReadCover ----------

// ReadCover — 读取封面文件.
//  - path.basename 剥目录组件 + sibling-prefix 绕过防御 (与 ReadChapterTxt 同口径)
//  - 必须 === COVERS_DIR 或以 COVERS_DIR + sep 开头
func ReadCover(fileName string) ([]byte, error) {
        initStoragePaths()
        safe := filepath.Base(fileName)
        full := filepath.Join(coversDir, safe)
        if full != coversDir && !strings.HasPrefix(full, coversDir+string(filepath.Separator)) {
                return nil, nil
        }
        data, err := os.ReadFile(full)
        if err != nil {
                if os.IsNotExist(err) {
                        return nil, nil
                }
                return nil, err
        }
        return data, nil
}

// ---------- DownloadTxtWriter ----------

// DownloadTxtWriter — 流式写入器 (万章书不 OOM).
//  - Open: 打开 filePath ('w' 模式, 截断)
//  - Write: 逐段 append
//  - Finish: close + stat, 返回 {rel, size}
//  - Abort: close + 删除半成品 (失败即无文件卫生语义)
type DownloadTxtWriter interface {
        Rel() string
        Write(ctx context.Context, chunk string) error
        Finish(ctx context.Context) (rel string, size int64, err error)
        Abort(ctx context.Context) error
}

type downloadTxtWriter struct {
        file     *os.File
        filePath string
        rel      string
}

// downloadTxtTarget — 下载成品文件名计算 (清洗控制字符 + 按码点截断防超长).
func downloadTxtTarget(name string) (filePath, rel, fileName string) {
        cleaned := chapterSlugRe.ReplaceAllString(name, "_")
        runes := []rune(cleaned)
        if len(runes) > 100 {
                runes = runes[:100]
        }
        base := string(runes)
        runes2 := []rune(base)
        if len(runes2) > 80 {
                runes2 = runes2[:80]
        }
        fileName = string(runes2) + ".txt"
        filePath = filepath.Join(downloadsDir, fileName)
        rel = "downloads/" + fileName
        return
}

// OpenDownloadTxtWriter — 打开下载成品流式写入器.
func OpenDownloadTxtWriter(name string) (DownloadTxtWriter, error) {
        if err := EnsureDirs(); err != nil {
                return nil, err
        }
        fp, rel, _ := downloadTxtTarget(name)
        f, err := os.OpenFile(fp, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
        if err != nil {
                return nil, err
        }
        return &downloadTxtWriter{file: f, filePath: fp, rel: rel}, nil
}

func (w *downloadTxtWriter) Rel() string { return w.rel }

func (w *downloadTxtWriter) Write(ctx context.Context, chunk string) error {
        if chunk == "" {
                return nil
        }
        // ctx 取消检查 (写盘是阻塞 IO, 写入前检查)
        select {
        case <-ctx.Done():
                return ctx.Err()
        default:
        }
        _, err := io.WriteString(w.file, chunk)
        return err
}

func (w *downloadTxtWriter) Finish(ctx context.Context) (string, int64, error) {
        if err := w.file.Close(); err != nil {
                return "", 0, err
        }
        info, err := os.Stat(w.filePath)
        if err != nil {
                return "", 0, err
        }
        return w.rel, info.Size(), nil
}

func (w *downloadTxtWriter) Abort(ctx context.Context) error {
        if err := w.file.Close(); err != nil && !os.IsExist(err) {
                // ignore 已关闭
        }
        return os.Remove(w.filePath)
}
